prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -0,0 +1,136 @@
+
+    :param object value: value to be coerced to float, if possible.
+    :param object default: value to be returned if float conversion isn't possible.
+    :return: (*float/object*) -- float or default value as applicable.
+    """
+    try:
+        if isinstance(value, str):
+            value = value.replace(",", "")
+        return float(value)
+    except Exception:
+        return default
+
+
+def map_generator_to_sub_by_location(generator, substation_groupby):
+    """Determine a likely substation for a generator to be connected to. Priority order
+    of mapping is: 1) if location is available and one or more substations exist in that
+    ZIP code, map by location to closest substation within that ZIP code, 2) if location
+    is available but no substations exist in that ZIP code, map to the closest
+    substation within neighboring ZIP codes, 3) if only ZIP code is available
+    (no location), and one or more substations exist, map to an arbitrarily chosen
+    substation within that ZIP code, 4) if only ZIP code is available (no location)
+    but no substations exist in that ZIP code, return NA.
+
+    :param pandas.Series generator: one generating unit from data frame.
+    :param pandas.GroupBy substation_groupby: data frame of substations, grouped by
+        (interconnect, ZIP).
+    :return: (*int/pd.NA*) -- substation ID if the generator can be mapped successfully
+        to a substation, else pd.NA.
+    """
+    lookup_params = tuple(generator.loc[["interconnect", "ZIP"]])
+    if pd.isna(generator["lat"]) or pd.isna(generator["lon"]):
+        # No location available
+        try:
+            matching_subs = substation_groupby.get_group(lookup_params)
+            return matching_subs.index[0]
+        except KeyError:
+            return pd.NA
+    try:
+        # This ZIP code contains substations, this block will execute successfully
+        matching_subs = substation_groupby.get_group(lookup_params)
+    except KeyError:
+        # If this ZIP code does not contain substations, this block will execute, and
+        # we select a set of 'nearby' substations
+        zip_range = [int(generator.loc["ZIP"]) + offset for offset in range(-100, 101)]
+        zip_range_strings = [str(z).rjust(5, "0") for z in zip_range]
+        try:
+            matching_subs = pd.concat(
+                [
+                    substation_groupby.get_group((generator.loc["interconnect"], z))
+                    for z in zip_range_strings
+                    if (generator.loc["interconnect"], z) in substation_groupby.groups
+                ]
+            )
+        except ValueError:
+            # If no matching subs within the given interconnection and ZIPs, give up
+            return pd.NA
+    distance_to_subs = matching_subs.apply(
+        lambda x: haversine((x.LATITUDE, x.LONGITUDE), (generator.lat, generator.lon)),
+        axis=1,
+    )
+    return distance_to_subs.idxmin()
```

Comment:
Do we use the interconnect of the substation at all in the above function? It seems that we could only use the ZIP of the substations

Reply:
I'm following the original design from the collaborators, and one of the considerations may have been to avoid mapping to a 'neighboring' ZIP code in a different interconnect (see the `zip_range` section).

Reply:
I see, it can happen that when we augment the ZIP coverage, some could be in a different interconnect. Are we sure that each generator has an interconnect assigned before grouping?

Reply:
Using the `[ZIP-100, ZIP+100]` as the set of neighboring ZIPs to look for substations in could probably be improved as well, since the numerical distance between two ZIPs is only a rough proxy for the geographical distance between two ZIPs. E.g. looking at [this map](https://www.mapshop.com/united-states-3-digit-zip-code-wall-map/) we see that ZIP codes 822xx in Wyoming are bordered by ZIP codes 820xx, 826xx, and 827xx, not 821xx or 823xx.

Reply:
http://www.geonames.org/export/web-services.html when we need to improve the method